Added UI tests

Igor-Arena opened the pull request:
1. Created LoginPageClass file with global variables in "pages" folder
2. Created Login file for login test in "ui_tests"
3. Created LoginNegativeTest for negative test in "ui_tests"

siarheima-papaya commented on playwright.config.ts:
```diff
@@ -10,7 +10,7 @@ import { defineConfig, devices } from '@playwright/test';
  */
 export default defineConfig({
-  testDir: './tests/api_tests_pet_store',
+  testDir: './tests/ui_tests',
```
You can use './tests' as testDir. If you want to run specific folder or file, use console command. Here example https://playwright.dev/docs/running-tests#run-specific-tests

siarheima-papaya commented on tests/api_tests_pet_store/delete.spec.ts:
```diff
@@ -1,7 +1,7 @@
 import axios from 'axios';
 import {faker, simpleFaker} from '@faker-js/faker';
-import { baseUrl } from '../global_settings/constant';
+import { baseUrl } from '../global_settings_pet_store/constant';
```
you can move constant.ts file in root directory (Demo-repo). And then use it for all places were needed. From my point of view it will make framework structure clear

Igor-Arena replied:
I moved folder "global_settings pet_store" in to root directory and left file constant like it is. Let this file will be for api pets tests. And for UI I have "pages" folder

siarheima-papaya commented on tests/pages/loginPageClass.ts:
```diff
@@ -0,0 +1,21 @@
+import {Page} from "@playwright/test";
+
+export class LoginPage {
```
Better when class and file names are equal

siarheima-papaya replied:
Move "Pages" folder in root directory. No necessary keep all folders inside "tests" directory

siarheima-papaya commented on tests/pages/loginPageClass.ts:
```diff
@@ -0,0 +1,21 @@
+    };
+
+    //Locators on login page
+    userNameTextbox = () => this.page.locator('[data-test="username"]');
```
Initialise all locators inside constructor. Here example https://playwright.dev/docs/pom#implementation

siarheima-papaya replied:
userNameInput

Igor-Arena replied:
for putting locators into constructor it needs new different implementation

Igor-Arena replied:
> Initialise all locators inside constructor. Here example https://playwright.dev/docs/pom#implementation

to move locators in constructor it needs new different implementation

siarheima-papaya commented on tests/ui_tests/login.spec.ts:
```diff
@@ -0,0 +1,20 @@
+}
+
+test('Login', async ({page}) => {
+const executeLoginPageClass = new LoginPage(page);
```
Please, use shortcuts to format code https://stackoverflow.com/questions/29973357/how-do-you-format-code-in-visual-studio-code-vscode

siarheima-papaya replied:
Just `const loginPage = new LoginPage(page);`

siarheima-papaya commented on tests/ui_tests/login.spec.ts:
```diff
@@ -0,0 +1,20 @@
+test('Login', async ({page}) => {
+const executeLoginPageClass = new LoginPage(page);
+await executeLoginPageClass.goto();
+await executeLoginPageClass.userNameTextbox().fill(user1.userName);
```
Move all interaction with locators to Page level (create in LoginPage methods fillUserNameInput(), fillPasswordInput() and etc). And then call methods in test (e.g. loginPage.fillUserNameInput(userName)

siarheima-papaya commented on tests/ui_tests/login.spec.ts:
```diff
@@ -0,0 +1,20 @@
+import {test, expect} from "@playwright/test"
+import {LoginPage} from "../pages/loginPageClass";
+
+const user1 = {
```
No reason to create object. You can pass in name and password as strings directly in test. Or create string variables userName, userPassword

siarheima-papaya commented on tests/pages/loginPageClass.ts:
```diff
@@ -0,0 +1,21 @@
+    };
+
+    public async goto() {
+      await this.page.goto('https://www.saucedemo.com/');
```
url can be moved to constants

siarheima-papaya commented on tests/ui_tests/loginNegativeTest.spec.ts:
```diff
@@ -0,0 +1,16 @@
+await executeLoginPageClass.passwordTextbox().fill(user2InvalidCreds.userPassword);
+await executeLoginPageClass.loginButton().click();
+await expect(page.getByText('Epic sadface: Username and password do not match any user in this service')).toBeVisible();
+});
```
remove blank files 
![Screenshot 2023-10-23 at 17 19 17](https://github.com/Igor-Arena/Demo-repo/assets/141643051/6b141e1f-4879-4861-bb17-392c8562c8b4)

Igor-Arena replied:
I dont see these blank files

siarheima-papaya commented on tests/ui_tests/loginNegativeTest.spec.ts:
```diff
@@ -0,0 +1,16 @@
+import {test, expect} from "@playwright/test"
```
You can merge files for login tests.

siarheima-papaya commented on tests/ui_tests/login.spec.ts:
```diff
@@ -0,0 +1,20 @@
+await executeLoginPageClass.userNameTextbox().fill(user1.userName);
+await executeLoginPageClass.passwordTextbox().fill(user1.userPassword);
+await executeLoginPageClass.loginButton().click();
+await expect(page.getByText('Products')).toBeVisible();
```
create function on Page level isDashboardPageLoaded() in DashboardPage (create new file)

Igor-Arena replied:
can't understand what to do and why to do this

siarheima-papaya commented on tests/ui_tests/login.spec.ts:
```diff
@@ -0,0 +1,20 @@
+await expect(page.getByText('Products')).toBeVisible();
+await executeLoginPageClass.burgerMenuButton().click();
+await executeLoginPageClass.logoutButton().click();
+await expect(page.getByText('Login')).toBeVisible();
```
+

siarheima-papaya commented on tests/pages/loginPageClass.ts:
```diff
@@ -0,0 +1,21 @@
+
+    //Locators on login page
+    userNameTextbox = () => this.page.locator('[data-test="username"]');
+    passwordTextbox = () => this.page.locator('[data-test="password"]');
```
passwordInput

siarheima-papaya commented on tests/ui_tests/loginNegativeTest.spec.ts:
```diff
@@ -0,0 +1,16 @@
+import {test, expect} from "@playwright/test"
+import {LoginPage} from "../pages/loginPageClass";
+
+const user2InvalidCreds = {
```
+

siarheima-papaya commented on tests/ui_tests/loginNegativeTest.spec.ts:
```diff
@@ -0,0 +1,16 @@
+await executeLoginPageClass.userNameTextbox().fill(user2InvalidCreds.userName);
+await executeLoginPageClass.passwordTextbox().fill(user2InvalidCreds.userPassword);
+await executeLoginPageClass.loginButton().click();
+await expect(page.getByText('Epic sadface: Username and password do not match any user in this service')).toBeVisible();
```
create function getLoginErrorText() on Page level and then compare text

siarheima-papaya commented on pages/dashboardPage.ts:
```diff
@@ -0,0 +1,21 @@
+  logoutButton: Locator;
+  constructor(page) {
+    this.page = page;
+    this.burgerMenuButton = page.locator("//button[@id='react-burger-menu-btn']");
```
It's not a mistake. Just keep in mind that you can use `[id='react-burger-menu-btn']` to identify locator

siarheima-papaya commented on pages/dashboardPage.ts:
```diff
@@ -0,0 +1,21 @@
+  constructor(page) {
+    this.page = page;
+    this.burgerMenuButton = page.locator("//button[@id='react-burger-menu-btn']");
+    this.logoutButton = page.locator("//a[@id='logout_sidebar_link']");
```
The same. You can use `[id='logout_sidebar_link']`

siarheima-papaya commented on pages/loginPage.ts:
```diff
@@ -0,0 +1,54 @@
+  }
+
+  async getLoginErrorText () {
+    await expect(this.loginErrorText).toBeVisible();
```
All assertions should be in the Test level. No reason to wrap assertion in additional function

siarheima-papaya commented on pages/loginPage.ts:
```diff
@@ -0,0 +1,54 @@
+    await expect(this.loginErrorText).toBeVisible();
+  }
+
+  async logInValidCredentials(
```
Rename to login(). And then you can use valid and invalid credentials

siarheima-papaya commented on pages/loginPage.ts:
```diff
@@ -0,0 +1,54 @@
+  }
+
+  async logInValidCredentials(
+    validUserNameValue: string,
```
userName

siarheima-papaya commented on pages/loginPage.ts:
```diff
@@ -0,0 +1,54 @@
+
+  async logInValidCredentials(
+    validUserNameValue: string,
+    validUserPasswordValue: any
```
userPassword

siarheima-papaya commented on pages/loginPage.ts:
```diff
@@ -0,0 +1,54 @@
+    validUserNameValue: string,
+    validUserPasswordValue: any
+  ) {
+    await this.goto();
```
remove goto from login function. It's better to call goto() in Test level
E.g.
```
test() => {
loginPage.goto()
loginPage.login(userName, userPassword)
}
```

siarheima-papaya commented on pages/loginPage.ts:
```diff
@@ -0,0 +1,54 @@
+    await this.clickLoginButton();
+  }
+
+  //Error messages
```
Do you need this comment?

siarheima-papaya commented on tests/ui_tests/login.spec.ts:
```diff
@@ -1,20 +1,18 @@
+  await loginPage.fillUserNameInput(user1.userName);
+  await loginPage.fillPasswordInput(user1.userPassword);
+  await loginPage.clickLoginButton();
+  await expect(page.getByText("Products")).toBeVisible();
```
add function in dashBoardPage to verify is page was loaded. 
E.g.
```
  public async isPageLoaded) {
    boolean isPageLoaded = await this.title.isVisible;
    return isPageLoaded;
  }
```

siarheima-papaya commented on tests/ui_tests/login.spec.ts:
```diff
@@ -1,20 +1,18 @@
+  await expect(page.getByText("Products")).toBeVisible();
+  await dashboardPage.clickBurgerMenuButton();
+  await dashboardPage.clickLogoutButton();
+  await expect(page.getByText("Login")).toBeVisible();
```
+

siarheima-papaya commented on tests/ui_tests/loginNegativeTest.spec.ts:
```diff
@@ -1,16 +1,12 @@
+  await loginPage.fillUserNameInput(user2InvalidCreds.userName);
+  await loginPage.fillPasswordInput(user2InvalidCreds.userPassword);
+  await loginPage.clickLoginButton();
+  await loginPage.getLoginErrorText();
```
use assertion directly

siarheima-papaya commented on tests/ui_tests/loginNegativeTest.spec.ts:
```diff
@@ -1,16 +1,12 @@
-import {test, expect} from "@playwright/test"
-import {LoginPage} from "../pages/loginPageClass";
+import { test, expect } from "@playwright/test";
```
You can merge both scenarios (positive and negative) in one test file. And use beforeEach() hook to navigate to LoginPage

siarheima-papaya commented on constant.ts:
```diff
@@ -0,0 +1,12 @@
+export let baseApiUrl = "https://petstore.swagger.io/v2/pet/";
+export let baseUiUrl = "https://www.saucedemo.com/";
+
+export let user1 = {
```
let's call it "standardUserCredentials"

siarheima-papaya commented on tests/ui_tests/loginNegativeTest.spec.ts:
```diff
@@ -0,0 +1,13 @@
+import { LoginPage } from "../../pages/loginPage";
+import { user2InvalidCreds } from "../../constant";
+
+test("LoginNegativeTest", async ({ page }) => {
```
Let's provide more understandable test title. E.g. "User can't login with invalid credentials"

siarheima-papaya commented on tests/ui_tests/login.spec.ts:
```diff
@@ -0,0 +1,16 @@
+import { user1 } from "../../constant";
+
+
+test("Login", async ({ page }) => {
```
Let's rename test to "User can login with valid credentials"

Igor-Arena commented:
1. Rename test to "User can login with valid credentials"
2. Merged valid and invalid logins
3. Renamed test to "User can't login with invalid credentials"
4. Renamed "constant.ts" to "standardUserCredentials"

Igor-Arena commented:
Delete file with negative login test

siarheima-papaya commented on standardUserCredentials.ts:
```diff
@@ -0,0 +1,13 @@
+export let baseApiUrl = "https://petstore.swagger.io/v2/pet/";
```
Why do you need separate this separate file? Just use the existing `constant.ts` and add object
```
export let standardUserCredentials = {
  userName: "standard_user",
  userPassword: "secret_sauce",
};
```

siarheima-papaya commented on constant.ts:
```diff
@@ -0,0 +1,12 @@
+  userPassword: "secret_sauce",
+};
+
+export let user2InvalidCreds = {
```
We don't need this object. Let use faker.string.alpha() directly in tests

siarheima-papaya commented on pages/dashboardPage.ts:
```diff
@@ -0,0 +1,46 @@
+    await this.shoppingCartIcon.click();
+  }
+
+  //Assertions
```
It's not assetion. It's just method which return boolean. It could be used not only inside expect(). I think we can remove this comment

siarheima-papaya commented on pages/dashboardPage.ts:
```diff
@@ -0,0 +1,46 @@
+    await this.addToCartButton.click();
+  }
+
+async clickShoppingCartIcon () {
```
Could you please use one format style for files. Here no spaces before 'async', but they are before other 'async'

siarheima-papaya commented on pages/dashboardPage.ts:
```diff
@@ -0,0 +1,46 @@
+    let isLoaded = await this.titleProducts.isVisible();
+    return isLoaded;
+  };
+  
```
remove extra blank lines. Let's do everything with good practice :)

siarheima-papaya commented on tests/ui_tests/login.spec.ts:
```diff
@@ -0,0 +1,27 @@
+test("User can login with valid credentials", async ({ page }) => {
+  const loginPage = new LoginPage(page);
+  const dashboardPage = new DashboardPage(page);
+  const title = dashboardPage.isDashboardPageLoaded();
```
Also isDashboardPageLoaded() - async function. Add await and run test with your implementation. You will see that test failed

Igor-Arena commented:
1. Added before all/each to the login test
2. Renamed "title" to "isDashboardPageLoaded"
3. Moved title to correct step

siarheima-papaya commented on constant.ts:
```diff
@@ -0,0 +1,13 @@
+  userPassword: "secret_sauce",
+};
+
+export let clientInformation = {
```
No usage in the code. Please, double-check Do you need this object?

Igor-Arena replied:
this data came from my next stage, from buying item

Igor-Arena replied:
I will delete it from this branch

siarheima-papaya commented on tests/ui_tests/login.spec.ts:
```diff
@@ -0,0 +1,36 @@
+  await loginPage.fillUserNameInput(standardUserCredentials.userName);
+  await loginPage.fillPasswordInput(standardUserCredentials.userPassword);
+  await loginPage.clickLoginButton();
+  const isDashboardPageLoaded = dashboardPage.isDashboardPageLoaded();
```
**await** dashboardPage.isDashboardPageLoaded()

siarheima-papaya commented on tests/ui_tests/login.spec.ts:
```diff
@@ -0,0 +1,36 @@
+});
+
+test("User can't login with invalid credentials", async () => {
+  await loginPage.fillUserNameInput("TestName");
```
Please, use faker.string. alphanumeric(). It will generate a random string every test and provide more opportunities to find a potential bug

siarheima-papaya commented on tests/ui_tests/login.spec.ts:
```diff
@@ -0,0 +1,36 @@
+
+test("User can't login with invalid credentials", async () => {
+  await loginPage.fillUserNameInput("TestName");
+  await loginPage.fillPasswordInput("TestPassword");
```
The same. Use faker

Igor-Arena commented:
1. Added faker
2. Deleted unused variables from constant.ts

Igor-Arena commented:
1. Puted faker directly into function
